KAFKA-5902: Added sink task metrics

rhauch opened the pull request:
Added Connect metrics specific to source tasks, and builds upon #3864 and #3911 that have already been merged into `trunk`, and #3959 that has yet to be merged.

I'll rebase this PR when the latter is merged.

rhauch committed:
KAFKA-5902 Added sink task metrics

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -582,4 +626,161 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        private final Sensor offsetCompletion;
+        private final Sensor offsetCompletionSkip;
+        private final Sensor putBatchTime;
+        private final AtomicReference<Map<TopicPartition, OffsetAndMetadata>> consumedOffsets = new AtomicReference<>();
```
Do these need to be `AtomicReference`s? What are the threads accessing it? Seems like the main processing thread for the sink task would be the only one?

rhauch replied:
No, not anymore. I used to have the thread reading the metrics also accessing these and computing the lag on-the fly, but that wasn't correct. I'll remove the AtomicReferences.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -582,4 +626,161 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        public SinkTaskMetricsGroup(ConnectorTaskId id, ConnectMetrics connectMetrics) {
+            consumedOffsets.set(new HashMap<TopicPartition, OffsetAndMetadata>());
+            committedOffsets.set(new HashMap<TopicPartition, OffsetAndMetadata>());
+            metricGroup = connectMetrics.group("sink-tasks-metrics",
```
same question as other pr -- this is `sink-task-metrics` instead of `sink-tasks-metrics` in the KIP

rhauch replied:
yes. I'll correct.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -582,4 +626,161 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                    final OffsetAndMetadata committedOffsetMeta = committedOffsetEntry.getValue();
+                    long consumedOffset = consumedOffsetMeta.offset();
+                    long committedOffset = committedOffsetMeta.offset();
+                    maxLag = Math.max(maxLag, consumedOffset - committedOffset);
```
Not sure this is what people will generally mean by lag -- while the committed offset matters, normally if the consumer is in the process requesting the lag I think it'd mean the FetchRequest lag, i.e. how far behind *processing* the records is the consumer in comparison to what the broker indicates is the most recent offset.

in other words, I think i'd update this at the end of each `put()` and change from `committedOffsets` to `processedOffsets`.

rhauch replied:
Okay, that's easy to change. But just to be clear, there are two things that lag could represent:

1. how many records the sink task has yet to process relative to the most recent offset written to the topic partitions, or
1. how many records the sink task has yet to process relative to the most recent offset read by the consumer.

Once I move from committedOffsets to processedOffsets as you suggest, the lag will be the second definition. Is that the preferred definition of lag?

ewencp replied:
Hmm, good question. I may have actually been wrong about which values should be involved. I think @gwenshap and I had a long discussion about this awhile ago too and there are many ways you could define lag.

I think the real problem here is that we may not be exposing enough information from the consumer to compute what I would really think of as lag -- FetchRequests include high watermark info so you know how many records are in the log but not yet returned to you, and the consumer creates metrics based on that. But we don't have access to that info. A connector that commits on every message would look like it has 0 lag, but it could be very far behind in the topic.

rhauch replied:
It's probably not a good thing performance-wise to get the end offsets from the consumer each time we poll, right?

ewencp replied:
no, that'd be expensive as it is a network request per call. there might actually be a way to get at the info if it could be dug out of the metrics from the consumer, but that's obviously very much not ideal.

we actually might just need some info from users about what lag they think is important here too. they can get at the consumer lag themselves via a metric, so perhaps something connect-specific is fine. i just worry that they will see lag and assume it means essentially the same thing as it does for consumers. @gwenshap thoughts?

rhauch replied:
Yes, the consumer metrics do capture the lag. I did try to carefully write the description of the metric:

> The maximum lag in terms of number of records that the sink task is behind the consumer's position for any topic partitions.

Hopefully that is clear enough, but I agree that it'd be better to have user feedback. Perhaps the metric name `sink-record-lag-max` is not ideal?

rhauch committed:
KAFKA-5902 Changed to measuring task processing lag behind consumer

Changed the `sink-record-lag-max` metric to be the maximum lag in terms of number of records that the sink task is behind the consumer's position for any topic partitions. This is not ideal, since often “lag” is defined to represent how far behind the task (or consumer) is relative to the end of the topic partition. However, the most recent offset for the topic partition is not easy to access in Connect.

rhauch commented:
@ewencp I've updated the PR with a new commit that fixes the group name and changes the `sink-record-lag-max` metric to be:

> The maximum lag in terms of number of records that the sink task is behind the consumer's position for any topic partitions.

However, this still is not the "normal" definition of lag.

rhauch commented:
Oh, and this is rebased on `trunk` now, since #3959 has been merged.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -582,4 +625,159 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+
+            sinkRecordLag = metricGroup.metrics().sensor("sink-record-lag");
+            sinkRecordLag.add(metricGroup.metricName("sink-record-lag-max",
+                    "The maximum lag in terms of number of records that the sink task is behind the consumer's position for any " +
```
Ok, sorry, I'm thinking more about this now with review, and I guess this will always just be either 0 or 1 batch of messages since the processing -> put() will be synchronous for each batch collected from the consumer.

So I guess maybe the committed - consumed makes sense as it is the total still thought to be somewhere in flight (or more accurately, not yet known to be guaranteed delivered into the destination) does actually work. I think, as you mentioned, lag is just confusing there because you could be completely done processing, the data could be in the destination, and we may just not yet have gotten to a periodic commit yet. I mainly would worry about that since connect defaults don't commit all that frequently and it is hard to say what it means if, e.g., the HDFS connector returns a large "lag" since it *needs* large "lag" to write large files.

:( sorry, i think this might need some more thought

rhauch replied:
The number of records being _processed_ should always be less than or equal to the batch size; the worst case scenario is when the consumer has read the full batch from a single topic partition, and in this case the processing lag would be equal to the batch size.

However, with connectors like the HDFS connector where commit is a more critical stage of processing, the commit lag seems to be far more useful and important: how many records have been consumed but not yet _written completely_ (committed) to the sink system. 

I've got the changes locally to revert back to the _committed lag_, so just let me know what you prefer.

ewencp replied:
right, i may also be playing fast and loose w/ terminology in my responses :) But I think we're on the same page as to the meaning of the different options.

My concern with the HDFS uncommitted case is that compared to consumer lag where, at least roughly, you could pick a number and use it across all topics for alerting, this doesn't have that property. The value you would alert on for HDFS is completely different from the value you'd alert on for ES. In this way "lag" definitely seems inaccurate terminology -- the connector isn't really "behind", it just intentionally has uncommitted data.

I see a few options:

* Include both! Something like "processed-lag" and "committed-lag".
* Don't use "lag" when referring to uncommitted data. It could just be something like "uncommitted-count" (which then opens the question of doing it per partition, max across partitions, or sum over all partitions).

And of course these aren't necessarily mutually exclusive. The potential drawbacks I see with the first point is that maybe the processed lag just isn't that interesting since it is bounded by the # of messages that fit in a single fetch request anyway, and introducing both may lead to confusion if we're not super clear in the docs about which is more important to monitor. In particular, now that I've thought through it more I'm struggling to find a problem I could discover / diagnose based on the processed lag.

The drawback to the second is that we're diverging from the terminology used elsewhere. But maybe that's fine since, as we're discovering through this conversation, they are in fact different things.

You could always re-raise this in the KIP discussion thread to get additional feedback. Also, might be worth looking at whether streams has any relevant metrics and how they handled this. In some ways they have the same problem -- they have the actual consumer lag, but also the time for a message to make it through the topology and all downstream messages to be acked. At best I would guess they could track lineage within a subtopology and provide info based on that, but I am guessing those metrics may simply not be there due to the more complex processing graph that they have.

ewencp replied:
oh, one other thing I just thought of -- if you are filtering messages aggressively with an smt, will metrics like committed-lag become misleading since they may not advance very frequently?

rhauch replied:
Streams doesn't seem to track any kind of lag, and I would expect that can only be computed by comparing consumer and producer offsets. And that works because the streams are always writing back to topics.

Connectors, of course, have to track how many records are in-flight in the connector, before being written to Kafka (for source connectors) or after having read from Kafka (for sink connectors). It seems like what we want to measure, then, is the number of in-flight records, or IOW, the max & avg number of records that the connector is "holding" at any point in time.

So, how about the following new metrics for source connectors:
* `source-record-process-queue-count-max` - the maximum number of source records that the source task had created but had not yet been written to Kafka or discarded by the transformations
* `source-record-process-queue-count-avg` - the average number of source records that the source task had created but had not yet been written to Kafka or discarded by the transformations

and replacing the "lag" metrics we've been discussing with the following new metrics for sink connectors:
* `sink-record-process-queue-count-max` - the maximum number of sink records that have been read from Kafka and not yet transformed and processed by the the sink task
* `sink-record-process-queue-count-avg` - the average number of sink records that have been read from Kafka and not yet transformed and processed by the the sink task
* `sink-record-flush-queue-count-max` - the maximum number of sink records that have been processed but not yet flushed by the sink task
* `sink-record-flush-queue-count-avg` - the average number of sink records that have been processed but not yet flushed by the sink task

And regarding "lag", isn't it true that since the consumer offsets are committed _after_ the sink task has flushed the records that the normal consumer lag actually tells you how far behind the sink task is? Sure, it's a bit larger than actual lag for connectors like HDFS that don't really do anything on flush, but that seems a relatively smaller problem than the fact that the offsets are committed infrequently.

WDYT?

ewencp commented:
On the source side, it seems like the description better matches the `flush-queue` naming from the sink than `process-queue` as the remaining work would mostly be waiting for the producer to get the ack. The "processing" phase in both source and sink seems generally inconsequential, so I think it's really only useful including it if we want to detect things like extremely slow transformations/converters.

If we have 2 phases on one side, I guess I don't see why it wouldn't be symmetric and have 2 on the other. Waiting for the producer to get data to Kafka is equivalent to waiting for the sink connector to get data to the sync system.

> And regarding "lag", isn't it true that since the consumer offsets are committed after the sink task has flushed the records that the normal consumer lag actually tells you how far behind the sink task is?

The normal consumer lag is *fetch* lag, not committed. See how it is [computed here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java#L322-L328) (note besides `position`, there is a `committed` field in the `TopicPartitionState`)

rhauch committed:
KAFKA-5902 Removed lag metric, added “active record count” metrics

The lag metric didn’t correspond to the conventional use of “lag”, and thus was removed until we can read the consumer lag to compute a real “lag” for the sink task. Instead, we can measuer the number of “active records” that the source and sink tasks have “in-flight”.

rhauch commented:
After discussing with @ewencp, we came to the conclusion that we eventually want to measure "lag" of the sink task but we don't have enough information from the consumer to actually compute the "lag" that users would expect. We'll leave that for future implementation, as it still will be a very important metric.

What we have been trying to measure in this PR and as discussed above is actually the number of records that the sink (and source) tasks have "in-flight". Of course, the "in-flight" term is already used in existing metrics, and this isn't quite the same and we need to pick a different name. So, these are the new metrics that the latest commit adds for *source tasks*:

* `source-record-active-count`: The number of records that have been produced by this task but not yet completely written to Kafka.
* `source-record-active-count-max`: The maximum number of records that have been produced by this task but not yet completely written to Kafka.
* `source-record-active-count-avg`: The average number of records that have been produced by this task but not yet completely written to Kafka.

and these for the *sink task*:

* `sink-record-active-count`: The number of records that have been read from Kafka but not yet completely committed/flushed/acknowledged by the sink task.
* `sink-record-active-count-max`: The maximum number of records that have been read from Kafka but not yet completely committed/flushed/acknowledged by the sink task.
* `sink-record-active-count-avg`: The average number of records that have been read from Kafka but not yet completely committed/flushed/acknowledged by the sink task.

@ewencp please have a look. If there is already a `1.0.0` branch, we'll have to merge this onto `trunk` and `1.0.0`.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -582,4 +625,170 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                    activeRecords += diff;
+                }
+            }
+            activeRecords = Math.max(activeRecords, 0L);
```
should we just do this per partition? if there were any negative values, this could still give incorrect information.

rhauch replied:
I'm not sure it's valuable per-partition. How can you get negative numbers? Can offsets be negative? If so, wouldn't it suffice to add the absolute value of `diff`?

ewencp replied:
sorry, that was unclear. i meant put the `Math.max` inside the loop and check `diff` instead of the aggregated `activeRecords`. you shouldn't get negative numbers, but if a connector does something wonky with the offset commits, it could happen that committed offsets are ahead of consumed offsets.

rhauch replied:
Ack

rhauch committed:
KAFKA-5902 Changed algorithm to compute active record counts

The change was made to prevent anomolous active record counts if the connector task implementations returns incorrect offsets.

rhauch commented:
@ewencp Corrected the active record count algorithm as requested.

ewencp reviewed (approved):
LGTM
